src/Control/Monad/Rec/Class.purs

```diff
@@ -139,3 +141,14 @@ instance monadRecMaybe :: MonadRec Maybe where
 -- | ```
 forever :: forall m a b. MonadRec m => m a -> m b
 forever ma = tailRecM (\u -> Loop u <$ ma) unit
+
+
+-- | While supplied computation evaluates to `Just _`, it will be
+-- | executed repeatedly and results will be combined using monoid instance.
+whileJust :: forall a m. Monoid a => MonadRec m => m (Maybe a) -> m a
+whileJust m = tailRecM (\v -> m <#> maybe (Done v) (Loop <<< (v <> _))) mempty
```

Comment:
Can we expand this out into actual pattern matches? `maybe` is strict this is going to allocate unused objects all over the place. Since this is pretty low-level, and presumably will be used as a basis for many other combinators, I think it makes sense to take that into account.

Reply:
good point